### DISCUSSION TITLE
Support for CharacterVirtual

### DISCUSSION THREAD

**Discussion by vensauro** [godot-jolt/godot-jolt#631] [General]
On this comment https://github.com/godotengine/godot-proposals/issues/2751#issuecomment-1649733609 jolt have been cited, to be having a stair walk behaviour, digging a little i have seen that there are some [configs about this on the jolt character](https://github.com/jrouwe/JoltPhysics/blob/b1cd84951022a565a23f33ff77ffee9e71487bbc/Jolt/Physics/Character/CharacterVirtual.cpp#L1149).
So is possible to utilize this character in some future?

**Comment by vensauro** [comment]
I don't know to much of how things are going here, seeing a little the code, seems that are not using the jolt character right?  
, and the godot native character node are using the binded physics functions?

edit: nvm, i have seen a answer about how the `move_and_slide` methods are and how it works, i understand more about this now

**Reply by mihe** [reply]
This extension does not use Jolt's `CharacterVirtual` or `Character` in any way, no.

`move_and_slide` (which makes up the majority of `CharacterBody3D`) is largely agnostic to the underlying physics engine used in Godot. The part that's implemented by the physics implementation is the [`body_test_motion`](https://docs.godotengine.org/en/stable/classes/class_physicsserver3d.html#class-physicsserver3d-method-body-test-motion) method found in `PhysicsServer3D`. This method effectively runs a three-stage process of depenetrating the body, casting the body in the movement direction and then running a collision check where the cast ended up. This is then used to implement the actual logic behind `move_and_slide`, which happens on the Godot side of things, not in this extension.

I've had other people ask about exposing Jolt's own character controller, but at the moment I don't have any concrete plans to do so. I touched on a similar thing in #623 as well, but frankly I would rather see that Godot's own character controller be improved than try to "sideload" features through this extension and thereby diverge quite heavily in terms of its interface.

If you're interested in stair-stepping specifically, you have an example [here](https://github.com/Griiimon/CharacterBody-On-Stairs) of how to use `SeparationRayShape3D` to achieve such an effect. It's not perfect, but it might do the trick depending on your circumstances.

**Reply by vensauro** [reply]
thanks for the response @mihe, realy is very good being in par with the godot community philosophy and evolution
